### PR TITLE
feat(utilities): add `objectEntries` function

### LINKED ISSUE #455: request: add `objectEntries` (a strictly typed `Object.entries`) to `@sapphire/utilities`
### Is there an existing issue or pull request for this?

- [X] I have searched the existing issues and pull requests

### Feature description

The types built into TypeScript for `Object.entries` intentionally aren't fully strictly typed, however there are instances where a strictly typed `Object.entries` can be useful. This is where a custom `objectEntries` can come in.

### Desired solution

Add a function like this and export it:

```ts
export function objectEntries<T extends NonNullObject>(obj: T): [keyof T, T[keyof T]][] {
  return Object.entries(obj) as [keyof T, T[keyof T]][];
}
```

Then implement unit tests for this

### Alternatives considered

N.A.

### Additional context

Porting this code from my work tbh

### PR DESCRIPTION
**Pull request by RealShadowNova**
Resolves #455

### CODE REVIEW

File: packages/utilities/tests/objectEntries.test.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+describe('objectEntries', () => {
+	test('GIVEN basic THEN returns expected', () => {
+		const source = { a: 'Hello', b: 420 };
+		const expected = [
+			['a', 'Hello'],
+			['b', 420]
+		] as [string, unknown][];
+
+		expect(objectEntries(source)).toEqual(expected);
```
Comment (kyranet) [inline packages/utilities/tests/objectEntries.test.ts:11]: Same as the other PR, please assert the input type + test readonly objects.

**Review by kyranet — CHANGES_REQUESTED** [review]

File: packages/utilities/src/lib/objectEntries.ts [hunk 2]
```diff
@@ -0,0 +1,5 @@
+import type { NonNullObject } from './utilityTypes';
+
+export function objectEntries<T extends NonNullObject>(obj: T): [keyof T, T[keyof T]][] {
+	return Object.entries(obj) as [keyof T, T[keyof T]][];
```
Comment (imranbarbhuiya) [inline packages/utilities/src/lib/objectEntries.ts:4]: same as the other 2 prs. You can better make 2 generic typings in `utility-types` files which can give all keys type and all values type. And use that in these 3 functions

**Review by imranbarbhuiya — CHANGES_REQUESTED** [review]

**Comment by RealShadowNova** [comment]
Should I change the return type for the key to a `string` or to a template literal containing `number` in the case of an array?
image

**Review by favna — COMMENTED** [review]
template literal same as the other pr

**Review by favna — APPROVED** [review]

**Review by vladfrangu — DISMISSED** [review]

File: packages/utilities/src/lib/objectEntries.ts [hunk 3]
```diff
@@ -0,0 +1,5 @@
+import type { NonNullObject } from './utilityTypes';
+
+export function objectEntries<T extends NonNullObject>(obj: T): T extends ArrayLike<any> ? [`${number}`, T[number]][] : [keyof T, T[keyof T]][] {
+	return Object.entries(obj) as T extends ArrayLike<any> ? [`${number}`, T[number]][] : [keyof T, T[keyof T]][];
```
Comment (favna) [inline packages/utilities/src/lib/objectEntries.ts:4]: comment on #474 is a good one

```suggestion
	return Object.entries(obj) as T extends ArrayLike<infer Values> ? Values[] : [keyof T, T[keyof T]][];
```

image
image

**Review by favna — CHANGES_REQUESTED** [review]

**Pull request merged**